### PR TITLE
refactor: shared user types

### LINKED ISSUE #4774: Move duplicated types to the shared types definitions
Bruce added shared types in #4769. We now need to comb through the types and make sure to consolidate all duplicated types into the shared types file.

### LINKED ISSUE #4786: Add API response types to the frontend
This is quite a big issue, if you wish to work on it, lets split it into parts:

 - [ ] users
 - [x] configs
 - [x] results 
 - [x] psas
 - [x] quotes [zaitsev-oleksii](https://github.com/zaitsev-oleksii)
 - [x] leaderboards
 - [x] presets
 - [x] publicStats
 - [x] apeKeys
 - [x] configuration

### PR DESCRIPTION
**Pull request by Miodec**
### Description

Last one baby.

Closes #4774.
Closes #4786.

### CODE REVIEW

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/7953074475

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/7960675375

File: shared-types/types.d.ts [hunk 1]
```diff
@@ -459,4 +459,142 @@ declare namespace SharedTypes {
...
+    hourOffset?: number;
+  };
+
+  type UserTag = {
+    _id: string;
+    name: string;
+    personalBests: PersonalBests;
+  };
+
+  type UserProfileDetails = {
```
Comment (fehmer) [inline shared-types/types.d.ts:476]: not optional? previous definition
```typescript
type UserProfileDetails = {
    bio?: string;
    keyboard?: string;
    socialProfiles: {
      twitter?: string;
      github?: string;
      website?: string;
    };
  };
```

File: shared-types/types.d.ts [hunk 2]
```diff
@@ -459,4 +459,142 @@ declare namespace SharedTypes {
...
+    premium?: PremiumInfo;
+    quoteRatings?: UserQuoteRatings;
+    favoriteQuotes?: Record<string, string[]>;
+    lbMemory?: UserLbMemory;
+    inventory?: UserInventory;
+    banned?: boolean;
+    lbOptOut?: boolean;
+    verified?: boolean;
+    needsToChangeName?: boolean;
+    quoteMod?: boolean | string;
```
Comment (fehmer) [inline shared-types/types.d.ts:536]: what is the "string" used for?

Comment (Miodec) [reply shared-types/types.d.ts:536]: true is moderate everything, string is moderate a given language

File: backend/src/types/types.d.ts [hunk 3]
```diff
@@ -18,103 +18,66 @@ declare namespace MonkeyTypes {
...
-    id: number;
-    selected?: boolean;
-  };
-
-  type UserQuoteRatings = Record<string, Record<string, number>>;
+  type DBCustomTheme = WithObjectId<SharedTypes.CustomTheme>;
+
+  type DBUserTag = WithObjectId<SharedTypes.UserTag>;
+
+  // type User = {
```
Comment (fehmer) [inline backend/src/types/types.d.ts:42]: remove?!

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/7961550820

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/7962063154

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/7962063154

**Pull request merged**